Fix brute force mesh intersection function

hultonha opened the pull request:
Update: PR now comes with a test and is ready for review. The problem was we were not looking at the right 'view' of the mesh data when there is more than one `ModelLodAsset::Mesh`.

### Before:

https://user-images.githubusercontent.com/82228511/140947276-8d20518a-2975-41e6-948a-fb91883e5987.mp4

### After:

https://user-images.githubusercontent.com/82228511/140948367-a0bec721-6c15-432e-989c-38f36c3414cd.mp4

## Example Mesh

### Part 1

![image](https://user-images.githubusercontent.com/82228511/140948646-21715a5d-c952-4a84-bf4a-f82829027f29.png)

### Part 2

![image](https://user-images.githubusercontent.com/82228511/140948721-13b9aa62-d678-4e9e-9d08-df8f347a1bb9.png)

## Update: Tests!

I've added a test that reproduces the old issue, see the below output with my changes reverted

```bash
[----------] 1 test from BruteForceMultiModelIntersectsFixture
[ RUN      ] BruteForceMultiModelIntersectsFixture.IntersectWithSecondMesh
D:/o3de-2/Gems/Atom/RPI/Code/Tests/Model/ModelTests.cpp(1437): error: Value of: m_mesh->GetModel()->LocalRayIntersectionAgainstModel( AZ::Vector3(15.0f, 0.0f, 5.0f), -AZ::Vector3::CreateAxisZ(10.0f), AllowBruteForce, t, normal)
Expected: is equal to true
  Actual: false (of type bool)
D:/o3de-2/Gems/Atom/RPI/Code/Tests/Model/ModelTests.cpp(1438): error: Value of: t
Expected: is approximately 0.5
  Actual: 0 (of type float)
[  FAILED  ] BruteForceMultiModelIntersectsFixture.IntersectWithSecondMesh (74 ms)
[----------] 1 test from BruteForceMultiModelIntersectsFixture (74 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (75 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] BruteForceMultiModelIntersectsFixture.IntersectWithSecondMesh
```

and with the change reintroduced:

```
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from BruteForceMultiModelIntersectsFixture
[ RUN      ] BruteForceMultiModelIntersectsFixture.IntersectWithSecondMesh
[       OK ] BruteForceMultiModelIntersectsFixture.IntersectWithSecondMesh (70 ms)
[----------] 1 test from BruteForceMultiModelIntersectsFixture (72 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (72 ms total)
[  PASSED  ] 1 test.
```

## Update 2:

Additional test to catch pointer offset mistake

```
[==========] Running 2 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 2 tests from BruteForceMultiModelIntersectsFixture
[ RUN      ] BruteForceMultiModelIntersectsFixture.RayIntersectsWithFirstSubMesh
[       OK ] BruteForceMultiModelIntersectsFixture.RayIntersectsWithFirstSubMesh (73 ms)
[ RUN      ] BruteForceMultiModelIntersectsFixture.RayIntersectsWithSecondSubMesh
[       OK ] BruteForceMultiModelIntersectsFixture.RayIntersectsWithSecondSubMesh (56 ms)
[----------] 2 tests from BruteForceMultiModelIntersectsFixture (131 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test case ran. (132 ms total)
[  PASSED  ] 2 tests.
```

pollend commented on Gems/Atom/RPI/Code/Source/RPI.Reflect/Model/ModelAsset.cpp:
```diff
@@ -261,13 +268,13 @@ namespace AZ
                     }
 
-                    const float* p = reinterpret_cast<const float*>(&positionRawBuffer[index0 * positionElementSize]);
+                    const float* p = &positionPtr[index0 * StepSize];
```
any reason why you need a temporary variable.
```
a.Set(const_cast<float*>(&positionPtr[index0 * StepSize]));
```

rgba16f replied:
Suggestion: turn p into a value and remove the need for the const_cast on the set.

invertednormal replied:
Looks like it was done this way for speed
> // faster than AZ::Vector3 c-tor

Really Vector3::Set ought to take a const array

hultonha replied:
I'll leave this as-is for now if that's cool, have the `p` pointer does make debugging easier too so it's useful for that (though perhaps having separate variables instead of reusing `p` might be nice...

hultonha replied:
Also agreed `Vector3` should to take a const pointer/array, that seems like an oversight 🙁

invertednormal reviewed (approved):
Looks like good changes - clearly just reading the buffers directly was wrong without getting the offset. We should also add an issue to fix later to validate the type of the index buffer and positions (indices could be 16 bit, positions could be half floats etc).

hultonha commented:
I've added a failing a test for this scenario now and made a couple of related improvements (`Vector3` fix for `Set` and switched to use `GetSemanticBufferAssetView` which does exactly what the code in `LocalRayIntersectionAgainstMesh` was doing itself 🙂)

FYI I'm going to try and tidy-up the test now, but wanted to get it posted to show it was at least working!

hultonha commented:
Reran the tests after the test refactor

// before
```
[ RUN      ] BruteForceMultiModelIntersectsFixture.IntersectWithSecondMesh
D:/o3de-2/Gems/Atom/RPI/Code/Tests/Model/ModelTests.cpp(1404): error: Value of: m_mesh->GetModel()->LocalRayIntersectionAgainstModel( AZ::Vector3(QuadOffsetX, 0.0f, 5.0f), -AZ::Vector3::CreateAxisZ(10.0f), AllowBruteForce, t, normal)
Expected: is equal to true
  Actual: false (of type bool)
D:/o3de-2/Gems/Atom/RPI/Code/Tests/Model/ModelTests.cpp(1405): error: Value of: t
Expected: is approximately 0.5
  Actual: 0 (of type float)
D:/o3de-2/Gems/Atom/RPI/Code/Tests/Model/ModelTests.cpp(1406): error: Value of: normal
Expected: is close (X: 0, Y: 0, Z: 1)
  Actual: (X: 1, Y: 1, Z: 1) (of type class AZ::Vector3)
[  FAILED  ] BruteForceMultiModelIntersectsFixture.IntersectWithSecondMesh (67 ms)
[----------] 1 test from BruteForceMultiModelIntersectsFixture (68 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (69 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] BruteForceMultiModelIntersectsFixture.IntersectWithSecondMesh
```

// after
```
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from BruteForceMultiModelIntersectsFixture
[ RUN      ] BruteForceMultiModelIntersectsFixture.IntersectWithSecondMesh
[       OK ] BruteForceMultiModelIntersectsFixture.IntersectWithSecondMesh (76 ms)
[----------] 1 test from BruteForceMultiModelIntersectsFixture (76 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (77 ms total)
[  PASSED  ] 1 test.
```

greerdv commented on Gems/Atom/RPI/Code/Source/RPI.Reflect/Model/ModelAsset.cpp:
```diff
@@ -248,8 +236,15 @@ namespace AZ
+                    positionRawBuffer.data() + (positionBufferViewDesc.m_elementOffset * positionBufferViewDesc.m_elementSize));
+
+                constexpr int StepSize = 3; // number of values per vertex (x, y, z)
+                for (uint32_t indexIter = 0; indexIter <= indexBufferViewDesc.m_elementCount - StepSize;
```
Wouldn't this be equivalent to using < rather than <= and losing the - StepSize? Seems like that would be slightly easier to read.

hultonha replied:
That's true, this preceded my change but I can update it, good call :+1:

greerdv commented on Gems/Atom/RPI/Code/Tests/Model/ModelTests.cpp:
```diff
@@ -1031,42 +1035,76 @@ namespace UnitTest
+            Add(lodCreator, positions, positionCount, positionOffset, positionBuffer, indices, indexCount, indexOffset, indexBuffer);
+        }
+
+        // complete the asset lot creation process
```
is "lot" correct here?

hultonha replied:
You're right it's not, this should be `load`, will fix :+1:

hultonha replied:
actually it should be `lod` 🤦‍♂️

burelc-amzn commented on Gems/Atom/RPI/Code/Tests/Model/ModelTests.cpp:
```diff
@@ -1306,4 +1344,71 @@ namespace UnitTest
+        EXPECT_THAT(
+            m_mesh->GetModel()->LocalRayIntersectionAgainstModel(
+                AZ::Vector3(QuadOffsetX, 0.0f, 5.0f), -AZ::Vector3::CreateAxisZ(10.0f), AllowBruteForce, t, normal),
+            testing::Eq(true));
```
Nit: You can use `testing::IsTrue()` here:
```suggestion
            testing::IsTrue());
```

hultonha replied:
Sure I can update that :+1:

burelc-amzn commented on Gems/Atom/RPI/Code/Tests/Model/ModelTests.cpp:
```diff
@@ -989,6 +990,9 @@ namespace UnitTest
         uint32_t{ 0 }, 2, 1, 1, 2, 3, 4, 5, 6, 5, 7, 6, 0, 4, 2, 4, 6, 2, 1, 3, 5, 5, 3, 7, 0, 1, 4, 4, 1, 5, 2, 6, 3, 6, 7, 3,
     };
 
+    static constexpr AZStd::array QuadPositions = { -1.0f, 1.0f, 0.0f, 1.0f, 1.0f, 0.0f, -1.0f, -1.0f, 0.0f, 1.0f, -1.0f, 0.0f };
+    static constexpr AZStd::array QuadIndices = { uint32_t{ 0 }, 2, 1, 1, 2, 3 };
+
```
Do these quads make a cube? I think these could use a better name.

hultonha replied:
Nope just a quad 🙂 I just wanted a valid mesh that was different to the Cube above

hultonha commented on Gems/Atom/RPI/Code/Tests/Model/ModelTests.cpp:
```diff
@@ -1031,42 +1035,76 @@ namespace UnitTest
+        {
             lodCreator.BeginMesh();
-            lodCreator.SetMeshAabb(AZ::Aabb::CreateFromMinMax({-1.0f, -1.0f, -0.5f}, {1.0f, 1.0f, 0.5f}));
+            lodCreator.SetMeshAabb(AZ::Aabb::CreateFromMinMax({ -1.0f, -1.0f, -0.5f }, { 1.0f, 1.0f, 0.5f }));
```
These tests don't reply on the AABB so don't go through this code path which is why the values aren't super important, I can update this but can you please remove request changes as otherwise I'll be blocked all tomorrow and am out on Friday to merge, I'd like to get this fix merged soon now.

amzn-tommy replied:
Looks like our model class does the aabb check (using the AABB from the ModelAsset) then calls LocalRayIntersectionAgainstModel. We should really just move that inside LocalRayIntersectionAgainstModel. No reason I can think of to skip the AABB check.

No need to fix it, I created a GHI to follow up on it https://github.com/o3de/o3de/issues/5523

hultonha commented on Gems/Atom/RPI/Code/Tests/Model/ModelTests.cpp:
```diff
@@ -1031,42 +1035,76 @@ namespace UnitTest
-                    AZ::RHI::BufferViewDescriptor::CreateStructured(0, static_cast<uint32_t>(indicesCount), sizeof(uint32_t))
-                });
-            }
+            AZStd::copy(indices, indices + indexCount, reinterpret_cast<uint32_t*>(const_cast<uint8_t*>(indexBuffer->GetBuffer().data())));
```
Hmm you could be right... but I'll have to check why the test passes and gets the right data... I'll double check tomorrow

amzn-tommy replied:
You're overwriting the first few vertices of the cube with the vertices from the quad. So the resulting model still has your offset quad, plus some vertices from the cube. So the intersection test will pass, since the quad is there.

hultonha replied:
You're quite right, good spot! This is now fixed and improved by this commit - https://github.com/o3de/o3de/pull/5447/commits/b6dfd9e0f68494dee0c2a9103906baea168a0666
